scripts: extract_dts_includes.py: fix multiple include in bindings

b0661 opened the pull request:
Correctly process multiple include files given to the
!include command of the YAML loader.

The fix only targets the sequential definition of include files.

The PR changes the 'node_type' node property to list type to account for multiple base types.

Fixes #7067

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>

b0661 committed:
scripts: extract_dts_includes.py: fix multiple include in bindings

Correctly process multiple include files given to the
!include command of the YAML loader.

The fix only targets the sequential definition of include files.

Fixes #7067

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>

tbursztyka commented on scripts/dts/extract_dts_includes.py:
```diff
@@ -584,15 +584,24 @@ def yaml_traverse_inherited(node):
+        if isinstance(node['inherits'], list):
+            inherits_list  = node['inherits']
+        else:
+            inherits_list  = [node['inherits'],]
```
isn't the ',' superfluous?

lpereira replied:
For lists, yes.  (It's not for tuples.)  In any case, it's harmless.

tbursztyka reviewed (commented):
I guess you can get rid of the "RFC" in the title as this is fixing an issue.

Try to recheck, CI failed for unrelated reasons to your pr.

b0661 commented:
@tbursztyka 

> I guess you can get rid of the "RFC" in the title as this is fixing an issue.

Yes it is fixing an issue but to do that it changes the 'node_type' node property to list type. I think @erwango introduced the 'node_type' property and he should decide whether a list is acceptable.

lpereira commented on scripts/dts/extract_dts_includes.py:
```diff
@@ -584,15 +584,24 @@ def yaml_traverse_inherited(node):
     :return: node
     """
 
+    if 'node_type' not in node.keys():
```
I don't know what `node` is, but if it's a dict, why use the `in` operator in the return of `keys`, instead of just using it in the `node` object directly?  Calling the `keys` method will allocate a new tuple with all the keys in the dict, and the `in` operator will then perform a linear search in that tuple; whereas, if using the `in` operator with `node` directly, a hash table lookup will be performed.

b0661 replied:
node is a dictionary. The .keys() thing is just taken from the original file. I will change.

b0661 commented:
@erwango 
I could not find a usage of the node['node_type'] property - so I removed the [RFC] label. Please have a look.

@lpereira, @tbursztyka 
Removed the superfluous .keys(). I stiil think the comma in [xxx,] is needed because in case xxx is a string you get a list  of characters otherwise.

tbursztyka reviewed (approved):
ok, fine by me

erwango reviewed (commented):
I'm ok about node_type, which is not used yet anyway.

One question about the global behavior when including multiple bindings in case of property overlap (if same property is included twice, at same level, via the multiple inclusion).
IIRC, until now there was a mechanism of overload (base value overloaded by node values). What would be the result when including multiple bindings? Establishing a list might not make sense in every cases.
I can see in #7067 that you are not facing this case, but maybe we should document a predefined behavior. Maybe even forbid it..

erwango commented:
> Documenting it is a good idea. Is there already a documentation "place" or do you mean in the code?

Throwing a meaningful error would be nice. Then `dts/bindings/device_node.yaml.template` should be a good candidate since there isn't a proper section in zephyr doc yet I think.

b0661 commented:
> Throwing a meaningful error would be nice. 

Tested it. Throwing an error does not work as the title and description property is set by all bindings. I will just inform about duplicates. This should be sufficent to  track unwanted behaviour.

> Then dts/bindings/device_node.yaml.template should be a good candidate

Ok. Will add.

b0661 commented:
@erwango 

Added warning message on duplicates and documentation.

erwango commented:
> Tested it. Throwing an error does not work as the title and description property is set by all bindings. I will just inform about duplicates. This should be sufficent to track unwanted behaviour.

What about tracking exception if property is different from 'title' or 'description'?
Also, I think that commenting that this should be avoided and that "first one prevails" is a bit contradictory. We're not being clear enough. Either this should not be used either this is a documented feature, it cannot be  both.
I'm not opposed to have this as a feature. Though if we're to propose it as a feature, I just want it to have a working and meaningful example. This can come later when the need arrives. 

Sorry to insist, I don't want to block anything but I've seen that people can get easily confused with this whole mechanism, and we should ensure people master what they're doing.
